lib/Dialect/Poly/IR/PolyAttributes.cpp

```diff
@@ -0,0 +1,180 @@
+#include "include/Dialect/Poly/IR/PolyAttributes.h"
+
+#include "llvm/include/llvm/ADT/SmallSet.h"      // from @llvm-project
+#include "llvm/include/llvm/ADT/StringExtras.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace poly {
+
+void PolynomialAttr::print(AsmPrinter &p) const {
+  p << '<';
+  p << getPolynomial();
+  p << '>';
+}
+
+/// Try to parse a monomial. If successful, populate the fields of the outparam
+/// `monomial` with the results, and the `variable` outparam with the parsed
+/// variable name.
+ParseResult parseMonomial(AsmParser &parser, Monomial &monomial,
+                          llvm::StringRef *variable, bool *isConstantTerm) {
+  APInt parsedCoeff(APINT_BIT_WIDTH, 0);
+  auto result = parser.parseOptionalInteger(parsedCoeff);
+  if (!result.has_value()) {
+    parsedCoeff = APInt(APINT_BIT_WIDTH, 1);
```

Comment:
nit/ my own understanding: since it's an optional integer to parse (so it doesn't set `parsedCoeff` to something unexpected when it doesn't encounter an integer), why not make 1 the default value and then only check if it has a failure value to emit error?

Reply:
Ah, yes, this was an artifact of me learning as I went. I initially declared `APInt parsedCoeff` with no initial value, but not specfiying the bit width (using the default bit width) causes assertion failures when it tries to compare two values with different bit widths. So then I added the initializer to fix that and didn't notice that the control flow could be simplified as a result. Will fix.